### ISSUE TITLE
A few thoughts on the paint editor

### ISSUE THREAD

**Issue by griffpatch** [LLK/scratch-paint#114]
I really love what you are doing with the editor and I feel scratchers will have a lot of fun with this.
I do note that Scratch 2 started out with a more powerful vector editor that was scaled back shortly after beta (or was it during beta?) because kids didn't understand the standard vector editing toolset. I understand why that was done, but it did make editing vectors rather tricky in Scratch 2.

So at present I live the inclusion of vector nodes with handles - These are sorely missed in Scratch 2 and I'd hate to see them go away. However, it will be tricky for a kid to know the difference between the select, and select node tools or the brush and pen tools. Are you intending to go down the route of it being a learning experience and perhaps provide a very simple tutorial / help system that explains each one? I think that would be awesome... perhaps even with linked inline video showing how to use them? - kids love that.

Colour palettes - I think you really need to add a nice big colour palette to the screen. A kid wants to click a nice bright color and then start painting :). The little color click boxes are possibly too small, and make selecting the same color again quite difficult. A color picker will go some way to fixing that, but then you'd need a way to apply that to an area afterwards (a fill tool again?). Another idea is to have a previously used color swatch list in the color palette popup (or could you even scan the SVG for used colors?)

Zoom (and pan) - I'll assume these are coming of course :)

Icons - A lot of work was put into designing the Scratch 3 blocks down to every little detail. At present the sprite palette (bottom right of the main scratch screen), the sound editor and paint editor all have quite different icon styles... some big and squishy, some small and very technical looking. I do like the look of each different style, but is the plan to have these all designed to keep to a specific look and feel?

Opacity / Transparency - It would be very handy to include a transparency slider to the color mixer, I've seen scratchers going to crazy lengths to get a semi opaque fill.

Rotate 90 degrees - Love the flip/mirror tool - Any change a rotate 90 degrees tool could sit alongside?

Combine - When I saw your Brush and Eraser tools I was well impressed. It did make me think that you could then consider adding a 'combine' tool that unions two shapes into one, or even a subtract. These sorts of tools are so very fun to play with and kids would have hours of creative fun with them. Plus they would actually be super useful.

I hope I didn't start to get too carried away there. I know if it was up to me then I'd probably get too excited and go overboard with features ;) - But I think that having layers of features (akin to how you now have the top category of scratch blocks) could work here too. The scratch community is so much more than just programming.

**Comment by thisandagain** [comment]
Thanks for the feedback! ❤️ 

Many of these are things that we are working on, but for some others as you note we try to keep a careful balance between making the experience as understandable (low floor) and flexible (wide walls) as possible. We will be testing this with kids starting next week and you'll likely see some significant changes as we gather feedback from users. We will try to track all of the features and testing notes in issues on this repo as publicly as we can.

**Comment by carljbowman** [comment]
Woo! Thanks for all the really thoughtful feedback. You certainly did not get too carried away, we are all very excited and eggar to continue to evolve the Vector Editor. At this point we are approaching the editor in phases. Each iteration will be adding / refining features and having kids try the vector editor in workshops. We are on iteration 2, which focuses on a few tools and a bit more detailed UI to help give state feedback.

**Re Vector Handles:** Including this level of vector manipulation has always been a tension for us. Impart for the reasons you described: there may be a conceptual confusion between brush and pen (as those follow more concrete understanding of drawing) vs curved point manipulation (which can be very abstract). We are curious to see how these will fair in workshops. This is one place we have fewer answers than others, but are actively thinking about ways of on-boarding. We are also trying to think of ways to help clarify between the Bitmap editor and the Vector editor.

**Re Color:** We would like to include a swatch list in the color picker. So there are both options of picking a color with the sliders, but also have some color swatches to choose from. We've also talked about the ability to save / add colors to the swatch list, but this may be a "down the road" feature. To your point about discoverability, we could also imagine having larger swatches appear else where in the editor for quick use. Right now there is a lot of attention focused on the state of color as you transition tools, objects, paths, and groups.

**Re Zoom and pan:** Of course! @paulkaplan and I just had an initial conversation about that yesterday.

**Re Icons:** Like the Vector Editor we have been working in iterations, but we will do another pass on the sound editor to sync the visual language in the icons with the rest of the UI.

**Re Transparency:** Certainly something we are considering, but as I mentioned before we are still working through some of the _statefulness_ of color. This would also have implications on the swatches. But none-the-less an exciting space to consider.

**Re Rotation:** OoOo we'll def will take that into consideration. 

**Re Union / Subtract:** We have been a bit hesitant to include this functionality. It was discussed but we tabled it as we would like to consider the UX a bit more.

Thanks again for your thoughts. We still have a ways to go both in the vector editor and the rest of 3.0.
